src/XrdNet/XrdNetUtils.cc

```diff
@@ -938,3 +940,72 @@ bool XrdNetUtils::Singleton(const char  *hSpec, const char **eText)
 //
    return isSingle;
 }
+
+bool XrdNetUtils::ConnectWithTimeout(int sockfd, const struct sockaddr* clientAddr,size_t clientAddrLen, uint32_t timeout_sec, std::stringstream &errMsg) {
+
+  if (!SetSockBlocking(sockfd, false, errMsg)) {  // Set to non-blocking
+    close(sockfd);
+    return false;
+  }
+
+  int result = connect(sockfd, clientAddr, clientAddrLen);
+  if (result == 0) {
+    //We've managed to connect immediately
+    SetSockBlocking(sockfd, true, errMsg);  // Set the socket back to blocking
+    return true;
+  } else if (errno != EINPROGRESS) {
+    errMsg << "Connection failed: " << strerror(errno);
+    close(sockfd);
+    return false;
+  }
+
+  struct pollfd fds;
+  fds.fd = sockfd;
+  fds.events = POLLOUT;  // Wait for the socket to be ready to write
+
+  result = poll(&fds, 1, timeout_sec * 1000); //Timeout in milliseconds
+
+  if (result < 0) {
+    errMsg << "Poll error: " << strerror(errno);
+    ::close(sockfd);
+    return false;
+  } else if (result == 0) {
+    errMsg << "Connection timed out";
+    ::close(sockfd);
+    return false;
+  } else {
+    if (fds.revents & POLLOUT) {
+      int so_error;
+      socklen_t len = sizeof(so_error);
+      getsockopt(sockfd, SOL_SOCKET, SO_ERROR, &so_error, &len);
+      if (so_error == 0) {
+        // We managed to connect
+        SetSockBlocking(sockfd, true, errMsg);  // Set back to blocking
+        return true;
+      } else {
+        errMsg << "Connection failed after poll: " << strerror(so_error);
+        ::close(sockfd);
+        return false;
+      }
+    }
+  }
```

Comment:
It's probably fine to merge this as is, but I'd consider refactoring this part to avoid nested `if`s, and such that the success case is when we arrive at the end with a `return true;`.

Reply:
Done with 773d2469aa4a34e05d6d2ba61b656a8b35303db1